src/graphics/device.js

```diff
@@ -2812,6 +2840,200 @@ Object.assign(pc, function () {
+                    this._shaderStats.vsCompiled++;
+                    // #endif
+                } else {
+                    this.fragmentShaderCache[src] = glShader;
+                    // #ifdef PROFILER
+                    this._shaderStats.fsCompiled++;
+                    // #endif
+                }
+            }
+
+            return glShader;
+        },
+
+        compileAndLinkShader: function (shader) {
+            var gl = this.gl;
+
+            var definition = shader.definition;
+            var glVertexShader = this.compileShaderSource(definition.vshader, true);
+            var glFragmentShader = this.compileShaderSource(definition.fshader, false);
+
+            var glProgram = gl.createProgram();
+
+            gl.attachShader(glProgram, glVertexShader);
+            gl.attachShader(glProgram, glFragmentShader);
+
+            if (this.webgl2 && definition.useTransformFeedback) {
+                // Collect all "out_" attributes and use them for output
+                var attrs = definition.attributes;
+                var outNames = [];
+                for (var attr in attrs) {
+                    if (attrs.hasOwnProperty(attr)) {
+                        outNames.push("out_" + attr);
+                    }
+                }
+                gl.transformFeedbackVaryings(glProgram, outNames, gl.INTERLEAVED_ATTRIBS);
+            }
+
+            gl.linkProgram(glProgram);
+
+            // Cache the WebGL objects on the shader
+            shader._glVertexShader = glVertexShader;
+            shader._glFragmentShader = glFragmentShader;
+            shader._glProgram = glProgram;
+
+            // #ifdef PROFILER
+            if (definition.tag === pc.SHADERTAG_MATERIAL) {
+                this._shaderStats.materialShaders++;
+            }
+            // #endif
+        },
+
+        createShader: function (shader) {
+            this.compileAndLinkShader(shader);
+
+            this.shaders.push(shader);
+        },
+
+        destroyShader: function (shader) {
+            var idx = this.shaders.indexOf(shader);
+            if (idx !== -1) {
+                this.shaders.splice(idx, 1);
+            }
+
+            if (shader._glProgram) {
+                this.gl.deleteProgram(shader._glProgram);
+                shader._glProgram = null;
+                this.removeShaderFromCache(shader);
+            }
+        },
+
+        postLink: function (shader) {
+            var gl = this.gl;
+
+            var program = shader._glProgram;
+            var vertexShader = shader._glVertexShader;
+            var fragmentShader = shader._glFragmentShader;
+
+            var definition = shader.definition;
+
+            // #ifdef PROFILER
+            var startTime = pc.now();
+            this.fire('shader:link:start', {
+                timestamp: startTime,
+                target: this
+            });
+            // #endif
+
+            // Check for errors
+            var addLineNumbers = function (src) {
```

Comment:
this is allocating a new function for every function, can this be a method instead?